### ISSUE TITLE
Json Request example

### ISSUE THREAD

**Issue by Art1k** [SmartBear/ready-api-testserver-client#13]
Hi,

I've been working on creating a recipe that will post parameterized json and i haven't found any actual example containing headers,parameters inside the body and how it's implemented with Code-based recipe approach.

Like adding header,mediatype,body parameters from other step \ dataSource

Thanks in advance,

**Comment by olensmar** [comment]
Hi!

Just to be sure I understand; are looking for how to expand datasource properties inside a POST/PUT message body or parameter value (for example) ?

/Ole

**Comment by Art1k** [comment]
Yes,actually it can be both.
For example:
...
"params": [{
		"javaClass": "java.util.HashMap",
		"map": {
			"registration.registrationTarget": 2,
			"person.firstName": "${DataSource#firstName}",
			"person.lastName": "${DataSource#lastName}",
			"person.phone1.num": "${DataSource#phone}",
			"person.email1.emailAddress": "${DataSource#email}",
...

**Comment by olensmar** [comment]
ok - so here it seems to me you want to send a HashMap serialized as json as a parameter value - is that correct?

**Comment by Art1k** [comment]
Yep,my requests can have several parts\maps but submitted as json

Full map example:
{
	"id": 0,
	"method": "***WebSiteService.*****",
	"params": [{
		"javaClass": "java.util.HashMap",
		"map": {
			"person.firstName": "${#[LeadSuite#Lead#DataSource]#firstName}",
			"person.lastName": "${#[LeadSuite#Lead#DataSource]#lastName}",
			"person.phone1.num": "${#[LeadSuite#Lead#DataSource]#phone}",
}
	}, {
		"javaClass": "java.util.HashMap",
		"map": {
			"PlatformType": 2,
			"AccountType": 0
		}
	}, {
		"javaClass": "java.util.HashMap",
		"map": {
			"hashedPassword": "${#TestSuite#hashedPassword}"
		}
	}]
}

**Comment by olensmar** [comment]
ok - so at the json-recipe level you refer to datasource properties with the traditional 
```
${<datasource name>#<property name>}
```
syntax, i.e. 
```json
{
  "testSteps": [
    {
      "type": "DataSource",
      "name": "The DataSource",
      "dataSource": {
        "grid": {
          "prop1": [
            "Madonna",
            "Tomten"
          ],
          "prop2": [
            "Kalle",
            "Pelle"
          ]
        }
      },
      "testSteps": [
        {
          "type": "REST Request",
          "method": "GET",
          "URI": "http://google.com/",
          "mediaType": "application/json",
          "parameters": [
            {
              "name": "q",
              "value": "${The DataSource#prop1}"
            },
            {
              "name": "usr",
              "value": "${The DataSource#prop2}"
            }
          ]
        }
      ]
    }
  ]
}
```
which should have corresponding methods in the java client. 

Regarding the creation of the JSON body from you java objects; are you doing that manually or using some kind of serialization? In either case you can assign the generated json to the request step with withRequestBody( String... ) method

does that help?

**Comment by Art1k** [comment]
So it should work like that?

.withRequestBody("{\n" +
                            "\t\"id\": 0,\n" +
                            "\t\"method\": \"com_tradefxl_crm_api_services_IWebSiteService.updateAccountAndAddProduct\",\n" +
                            "\t\"params\": [{\n" +
                            "\t\t\"javaClass\": \"java.util.HashMap\",\n" +
                            "\t\t\"map\": {\n" +
                            "\t\t\t\"person.firstName\": \"${#[MT4LeadSuite#Lead#DataSource]#firstName}\",\n" +
                            "\t\t\t\"person.lastName\": \"${#[MT4LeadSuite#Lead#DataSource]#lastName}\",\n" +
                            "\t\t\t\"person.phone1.num\": \"${#[MT4LeadSuite#Lead#DataSource]#phone}\",\n" +
                            "\t\t\t\"person.email1.emailAddress\": \"${#[MT4LeadSuite#Lead#DataSource]#email}\",\n" +
                            "\t\t\t\"person.country\": \"DE\",")
:

**Comment by olensmar** [comment]
yeah - except the datasource references would have to ref a teststep in your recipe. If you write your tests in groovy you could probably use the dynamic features of it to generate this block of json through code - or in java you could perhaps use a custom json serializer to make things easier!?

**Comment by Art1k** [comment]
Can i keep variables on Suite lvl?

**Comment by olensmar** [comment]
Unfortunately test recipes don't currently expose a Suite object - not sure what the best way to pass parameters between recipes would be (as discussed in the other issue...)...

**Comment by Art1k** [comment]
When i use this syntax inside my json it's not substituted i see it like that inside the request also

`"${Step1#firstName}"`

What could be the reason?

**Comment by spindelmanne** [comment]
It is substituted inside the TestServer, before submitting the request to the system under test.

**Comment by Art1k** [comment]
When i print my request, i see the following:

 "def request = testRunner.testCase.getTestStepByName("Step2").httpRequest.getRequestContent()

**Comment by Art1k** [comment]
I have to pass properties between step1 and others,so due to an issue with property transfer method i need to hold them on case level,but they're not substituted during request submition
